feat: #681 Stateful tabs

M123-dev opened the pull request.

M123-dev committed:
feat: first working stateful tab

M123-dev committed:
feat: enum instead of String

M123-dev commented:
@monsieurtanuki, I think I need some guidance here, so the tabs are now stateful, they kept rendered and get only hidden by the Offstage widget, but there is one problem, keeping the tabs rendered also keeps the camera open, this is of course not wanted behaviour but I am unsure how to get around this while keepin it stateful

M123-dev committed:
Update tab_navigator.dart

monsieurtanuki commented:
Hi @M123-dev!
The root problem comes from the fact that there were (and maybe still are) 3 `Widget`s (profile, scan, history) created at the same time, with the tabs just deciding which one to display.
Which is problematic with one `Widget` (scan) that needs the camera on, even when not displayed.

My quick suggestion:
* consider each tab as an actual page (`Scaffold`), and not a `Widget`
* don't create the pages at the same time - there's just one current page at a time
* if you leave the scan page, perhaps it's not enough to push a new page, camera-wise, because flutter may say "it's just the previous page, I keep the camera on, just in case we go back". Therefore you could either _explicitly_ close the camera when you leave the scan page, or pop the scan page first then push the new page (something like `Navigator.popAndPush` probably exist)

I hope this helps.

M123-dev committed:
feat: Working camera disable for other tabs

M123-dev committed:
fix: #640 stop camera when navigating further in the stack

M123-dev committed:
doc + null safety

M123-dev committed:
fix: activating scanner when rebuilding from ground up

M123-dev committed:
Updated goldens

The navbar got removed from the images because it is now not directly added by the pages anymore

M123-dev committed:
doc:

M123-dev commented:
Thanks for your answer @monsieurtanuki I decided to sticke with with closing the camera when I exit (other tab or continue navigating in the scanner) since poping and then pushing a new page wouldn't keep the state nor the next opend pages.
For the remaining points, we now do create all three pages at once where there is probably no good way around, but because of `WillPopScope` the tabs feel as if you have pushed them to the stack, as the back button switches back to the last tab. 

--- 

@gspencergoog
@Piinks thank you for your hint on the `AutomaticKeepAliveClientMixin` I have also tested this out but as you have already suspected this unfortunately did not work with a own navigator.

M123-dev commented:
### PullRequest summary:

- The navigation bar is now statefull, which means that it keeps not only things like the scroll position but also new opened pages.
- If you push to another page and then press the tab you are currently on you will be redirected back to the inital tab page.
- Because of WillPopScope the back buttons also work to get back to the inital tab.
- The scanner/camera gets deactivated when you switch to another tab or open any page (Search, Product, Ranking, ...) directly from the scanner.

monsieurtanuki commented on packages/smooth_app/lib/data_models/continuous_scan_model.dart:
```diff
@@ -40,6 +40,8 @@ class ContinuousScanModel with ChangeNotifier {
   final String languageCode;
   final String countryCode;
 
+  QRViewController? qrViewController;
```
Two things:
* (not sure) `late` would make sense
* private field would definitely make sense

Therefore:
* either `late QRViewController _qrViewController;`
* or `QRViewController? _qrViewController;`

M123-dev replied:
In dart there is no way to check if a late variable is initialized or not, so I am always a bit worried about to use them when it's unclear if it is. With the current behaviour it shouldn't happen since it doesn't try to restart the camera if it has not yet been build but it doesn't have to be alwys the case.
Made it private 👍🏼

monsieurtanuki commented on packages/smooth_app/lib/data_models/continuous_scan_model.dart:
```diff
@@ -96,8 +98,23 @@ class ContinuousScanModel with ChangeNotifier {
+  void restartQRView() {
+    if (qrViewController != null) {
+      qrViewController!.resumeCamera();
+    }
```
Not sure: would `_qrViewController?.resumeCamera();` work?

monsieurtanuki commented on packages/smooth_app/lib/data_models/continuous_scan_model.dart:
```diff
@@ -96,8 +98,23 @@ class ContinuousScanModel with ChangeNotifier {
+  void stopQRView() {
+    if (qrViewController != null) {
+      qrViewController!.stopCamera();
+    }
```
Not sure: would `_qrViewController?.stopCamera();` work?

monsieurtanuki commented on packages/smooth_app/lib/pages/scan/continuous_scan_page.dart:
```diff
@@ -4,27 +4,75 @@ import 'package:provider/provider.dart';
+  @override
+  void didPushNext() {
+    super.didPushNext();
+    if (model != null) {
+      model!.stopQRView();
+    }
```
Not sure: would `model?.stopQRView();` work?

M123-dev replied:
Somehow forgot about the `?` while writing this code, thanks for pointing out

monsieurtanuki commented on packages/smooth_app/lib/widgets/tab_navigator.dart:
```diff
@@ -0,0 +1,28 @@
+
+  @override
+  Widget build(BuildContext context) {
+    late Widget child;
+    if (tabItem == SmoothBottomNavigationTab.Profile) {
+      child = const UserPreferencesPage();
+    } else if (tabItem == SmoothBottomNavigationTab.History) {
+      child = const HistoryPage();
+    }
```
Difficult to understand why "scan" is not possible here.
At least put a comment.

M123-dev replied:
Done

monsieurtanuki commented on packages/smooth_app/lib/pages/smooth_bottom_navigation_bar.dart:
```diff
@@ -1,85 +1,109 @@
+    final bool offstage = _currentPage != tabItem;
+    return Offstage(
+      offstage: offstage,
+      child: tabItem != SmoothBottomNavigationTab.Scan
+          ? TabNavigator(
+              navigatorKey: _navigatorKeys[tabItem]!,
+              tabItem: tabItem,
+            )
+          : ScanPage(
+              offstage: offstage,
+              navigatorKey: _navigatorKeys[tabItem]!,
+            ),
```
This is confusing: why would the scan page be considered differently?

M123-dev replied:
Added a comment for that

monsieurtanuki commented on packages/smooth_app/lib/main.dart:
```diff
@@ -171,7 +171,7 @@ class SmoothAppGetLanguage extends StatelessWidget {
       languageCode,
     );
-    return SmoothBottomNavigationBar.getDefaultPage();
+    return SmoothBottomNavigationBar();
```
Then it's not a "bottom navigation bar": it's more a "page manager".

M123-dev replied:
Agree

monsieurtanuki commented on packages/smooth_app/lib/pages/scan/continuous_scan_page.dart:
```diff
@@ -4,27 +4,75 @@ import 'package:provider/provider.dart';
   Widget build(BuildContext context) {
+    if (!ModalRoute.of(context)!.isCurrent) {
+      isCurrent = false;
+      return const Center(child: Text('A unknown error occurred'));
```
A bit fishy.

monsieurtanuki replied:
I still think there's a flaw if you have to check twice, here and in ScanPage: `if (widget.offstage) ...`.

monsieurtanuki commented on packages/smooth_app/lib/pages/smooth_bottom_navigation_bar.dart:
```diff
@@ -1,85 +1,109 @@
+    SmoothBottomNavigationTab.Scan: GlobalKey<NavigatorState>(),
+    SmoothBottomNavigationTab.History: GlobalKey<NavigatorState>(),
   };
+  int _selectedIndex = 1;
```
What's the difference between `_currentPage` and `_selectedIndex`?

M123-dev replied:
There isn't a big difference, we only need it since the BottomNavigationBar needs a `currentIndex` int.
Of course we could run everything via the index, but I decided to use the `_currentPage` to keep it easier to read.

monsieurtanuki replied:
That's what I suspected. It's not easier to read `_selectTab(_pageKeys[index], index)` when you ask yourself: there are 2 parameters, that means something, perhaps it's possible to write `_selectTab(_pageKeys[0], 1)` in some strange cases.
I guess either `BottomNavigationTab get _currentPage => _pageKeys[_selectedIndex]` or the other way around would be easier to read.

M123-dev replied:
Updated it, now we just use the index of _currentPage instead of a own var

monsieurtanuki commented on packages/smooth_app/lib/pages/smooth_bottom_navigation_bar.dart:
```diff
@@ -1,85 +1,109 @@
-  static const List<SmoothBottomNavigationTab> _tabs =
-      <SmoothBottomNavigationTab>[
+class SmoothBottomNavigationBarState extends State<SmoothBottomNavigationBar> {
+  List<SmoothBottomNavigationTab> pageKeys = <SmoothBottomNavigationTab>[
```
Could still be `static const` and probably private, couldn't it?

monsieurtanuki reviewed (commented):
Hi @M123-dev!
To be honest I'm a bit lost; please have a look at my comments.

M123-dev committed:
Minor changes

monsieurtanuki commented on packages/smooth_app/lib/pages/scan/scan_page.dart:
```diff
@@ -39,9 +44,29 @@ class _ScanPageState extends State<ScanPage> {
+    //Don't build Scanner (+activate camera) when not on the Scan Tab
+    if (widget.offstage) {
+      _model!.stopQRView();
+      //shouldn't be visible since only build when offstage
+      return const Center(child: Text('A error occurred'));
```
I suggest:
```dart
return const Center(child: Text('shouldn\'t be visible since only build when offstage'));
```

M123-dev replied:
Sounds good

monsieurtanuki commented on packages/smooth_app/lib/widgets/tab_navigator.dart:
```diff
@@ -0,0 +1,30 @@
+
+  @override
+  Widget build(BuildContext context) {
+    late Widget child;
```
Here it's more `final` than `late`: we are 100% positive that the variable will be set.

monsieurtanuki commented on packages/smooth_app/lib/pages/page_manager.dart:
```diff
@@ -0,0 +1,113 @@
+          items: const <BottomNavigationBarItem>[
+            BottomNavigationBarItem(
+              icon: Icon(Icons.account_circle),
+              label: 'Profile',
```
Please add a `TODO` for translations.

monsieurtanuki commented on packages/smooth_app/lib/pages/page_manager.dart:
```diff
@@ -0,0 +1,113 @@
+  History,
+}
+
+class PageManager extends StatefulWidget {
```
I would add some comments here too, in order to explain the issue. Something like "Stateful tabs management".
Or link to the issue.

M123-dev committed:
fix: Updating product with multiple navigators

monsieurtanuki commented on packages/smooth_app/lib/pages/page_manager.dart:
```diff
@@ -0,0 +1,113 @@
+    // In order for the scanPage be to able to decide whether to activate the camera or not
+    // the offstage value has to be passed to it and can because of that not be
+    // handled by the TabNavigator
+    return Offstage(
+      offstage: offstage,
+      child: tabItem != BottomNavigationTab.Scan
+          ? TabNavigator(
+              navigatorKey: _navigatorKeys[tabItem]!,
+              tabItem: tabItem,
+            )
+          : ScanPage(
+              offstage: offstage,
+              navigatorKey: _navigatorKeys[tabItem]!,
+            ),
+    );
```
I suspect that you would just have to add the `offstage`parameter to `TabNavigator` in order not to deal with the special "scan" case.

M123-dev replied:
In some testing scenario it did not work but now it does, added it inside the TabNavigator eventhough it doesn't use the there build Navigator 👍🏼

monsieurtanuki commented on packages/smooth_app/lib/pages/scan/continuous_scan_page.dart:
```diff
@@ -4,27 +4,73 @@ import 'package:provider/provider.dart';
+      isCurrent = true;
+      //First building of scanner
+      setState(() {});
+    } else if (model != null) {
+      model!.restartQRView();
+    }
```
`model?.restartQRView();`

M123-dev committed:
Minor fixes

monsieurtanuki reviewed (commented):
Thank you @M123-dev for having taken into consideration most of my remarks.
There's still something I think is too complicated in the algorithm: of course the ScanPage is a bit different (with its onPause/onResume methods), but I do think it should be treated as the other 2 pages (with their empty onPause/onResume methods).
I also find the RouteObserver part a bit cumbersome, but I've never used it therefore I cannot give a relevant opinion on it.

M123-dev commented:
Hello @monsieurtanuki, thanks for your (as always) detailed review. 
I don't know if I follow you on the onPause/onResume methods. And to the RouteObserver, this basically has nothing to do directly with the StatefullTabs but fits to this PR. It fixes #640, since the camera doesn't go out when you push a new page over it. Maybe we could save a few lines there by just rebuilding the scanpage when poping to it and not looking first if we could just reactivate the camera but for maybe 5 lines that's not worth it. I'll try to add a few more comments there maybe this makes it clearer.

monsieurtanuki commented:
Hi @M123-dev!
`onPause`/`onResume` is a reference to Android java dev: those are callbacks that you typically use when the page appears/disappears and you want to reload/save data (cf. incoming phone call). That is similar to your camera on/off issue: we need the camera on if and only if the scan page is displayed.
About putting in the same PR a code that solves several issues, I'm not against that in general, but here we're talking about a PR where you started stating "I think I need some guidance here". Currently we already exchanged 33 posts for this PR, and you committed 11 times. If you can implement in this PR only the "stateful tabs" part (even with a stupid camera always on), and then in a next PR the camera on/off system, I'd be grateful, as a code reviewer.

M123-dev committed:
Simplification

M123-dev committed:
Simplification

M123-dev committed:
Simplification

M123-dev commented:
@monsieurtanuki I have done a bit of simplification, and documentation, I probably can't get this PR down to much less code

monsieurtanuki commented:
@M123-dev I'm currently in the `hive` code - I'll be able to review your code tomorrow at best.

M123-dev commented:
No problem @monsieurtanuki, I don't want to distract you from it, the PR's shouldn't conflict each other either.

monsieurtanuki committed:
Merge branch 'develop' into stateful-tabs

monsieurtanuki committed:
Update main.dart

monsieurtanuki committed:
Update product_list_page.dart

monsieurtanuki reviewed (approved):
Hi @M123-dev!
I don't understand everything around the Route Observer, but that's my problem, not yours :)
